feat: Add glibc-based static binary build support

crazywhalecc opened the pull request:
I have merged the upstream parts into the main branch and made some adjustments to build-static.sh. But I don't know much about the Docker distribution of FrankenPHP, maybe the Dockerfile needs some additional modifications.

This binary is base on gnu libc dynamically linked and other things are all built statically, this allows us to use it on almost all glibc distributions, while also enabling some features that pure static compilation does not have, such as `dl()`, etc.

Currently I only tested locally both on arm and x86 platform with `docker build -t gnu-static -f gnu-static.Dockerfile --build-arg FRANKENPHP_VERSION=1.0 .`

crazywhalecc commented on gnu-static.Dockerfile:
```diff
@@ -0,0 +1,134 @@
+FROM centos:7
```
We need to use old distro to build high compatibility binary. It uses glibc 2.17, and makes executable have lower glibc requirements. If we use gha runner directly, we will get binary that only available newer than that distro.

dunglas replied:
Ok, that makes sense!

DubbleClick commented:
@crazywhalecc would you mind giving me push access to your gnu-frankenphp repo?

crazywhalecc commented:
> @crazywhalecc would you mind giving me push access to this PR repo?

Added maintain access.

DubbleClick commented:
> > @crazywhalecc would you mind giving me push access to this PR repo?
> 
> Added maintain access.

PS C:\Users\m\source\repos\gnu-frankenphp> git push
remote: Permission to crazywhalecc/frankenphp.git denied to DubbleClick.
fatal: unable to access 'https://github.com/crazywhalecc/frankenphp/': The requested URL returned error: 403

crazywhalecc commented:
> > > @crazywhalecc would you mind giving me push access to this PR repo?
> > 
> > 
> > Added maintain access.
> 
> PS C:\Users\m\source\repos\gnu-frankenphp> git push remote: Permission to crazywhalecc/frankenphp.git denied to DubbleClick. fatal: unable to access 'https://github.com/crazywhalecc/frankenphp/': The requested URL returned error: 403

Ouch. I added you in [static-php/gnu-frankenphp](https://github.com/static-php/gnu-frankenphp) , forgot my account's repo, sorry, added.

DubbleClick commented:
I personally prefer spaces on git over tabs, but most of the file was using tabs, so I changed our added spaces to tabs instead.

DubbleClick commented:
looks good now other than the CHECKOV check for the Dockerfile. @crazywhalecc can you fix that?

dunglas commented:
Could you also add a GitHub Actions workflow to automatically publish the binary as we do for musl?

DubbleClick commented:
Yes, I'm still finishing up a few things related to this.

DubbleClick commented:
I'm a bit stumped. Just recompiled with bin/build-frankenphp, which does essentially the same thing, and it's working just fine.

DubbleClick commented:
seems related to FFI. I will have to investigate that further, because FFI works just fine when I compile spc and frankenphp from my host RHEL 9.5 machine.

DubbleClick commented:
I'm a bit concerned that the linux arm runner has been running for 5 hours, it seems to be insanely slow for some reason.

Resulting executable is ~220mb, ~80mb after UPX. The bulk of the size comes from -Wl,--export-dynamic, but it's the only way I managed to keep the symbols required to load extensions in the executable.

Nearly the same dockerfile also works with gcc14 on a almalinux 9 container, which would enable some nice optimisation improvements. I will later benchmark the two.

dunglas commented on .github/workflows/static.yaml:
```diff
@@ -176,12 +179,113 @@ jobs:
+      matrix:
+        platform: ${{ fromJson(needs.prepare.outputs.platforms) }}
+    name: Build ${{ matrix.platform }} GNU static binary
+    runs-on: ubuntu-24.04
```
```suggestion
    runs-on: ${{ startsWith(matrix.platform, 'linux/arm') && 'ubuntu-24.04-arm' || 'ubuntu-24.04' }}
```

This will run the ARM builds on ARM runners and speed up the builds.

DubbleClick replied:
ah, that explains it. thanks.

dunglas commented on gnu-static.Dockerfile:
```diff
@@ -0,0 +1,135 @@
+# syntax=docker/dockerfile:1
```
`gnu-static-builder` for consistency with `static-builder` (that we could rename `musl-static-builder` btw)?

DubbleClick replied:
good idea

DubbleClick replied:
I think static-builder-musl and static-builder-gnu would be better to keep them grouped together.

DubbleClick commented:
> Nearly the same dockerfile also works with gcc14 on a almalinux 9 container, which would enable some nice optimisation improvements. I will later benchmark the two.

Haha, gcc 14 isn't too happy with php source code. I patched a few sections and got it to build, doing performance testing against the gcc 10 (with -O3 -march native too) now.

dunglas commented:
We should wait for proper upstream support then, isn't it?

DubbleClick commented:
Yeah that's why I haven't committed anything for it. It would also require a more modern glibc version so it wouldn't make sense to distribute frankenphp with it.

DubbleClick commented:
Well, I can't make out any performance improvements by gcc 14 over gcc 10 on my old Ryzen 3700X. I'm guessing there may be some when running on a modern cpu, but I can't confirm that until tomorrow night.  Can't benchmark our production servers while they're in use :P.

Weirdly enough, I actually get slightly better response times out of -Os than -O3 for php's compilation.

Haven't tested different options for FrankenPHP itself because I would guess that you've benchmarked that. If anything we could try to use LTO but I suspect it wouldn't cause a major change.

DubbleClick commented:
Well I just spent the better part of two hours benchmarking various compilation settings and performance (and file size) are all very similar, to the point where I can't make any definitive statements. I think this is probably fine to be merged the way it is now.

dunglas commented on docker-bake.hcl:
```diff
@@ -143,3 +143,26 @@ target "static-builder" {
+        FRANKENPHP_VERSION = VERSION
+    }
+    secret = ["id=github-token,env=GITHUB_TOKEN"]
+}
```
```suggestion
}

```

dunglas commented:
Thank you very much for this @DubbleClick and @crazywhalecc. That's a huge step forward for the project.

DubbleClick commented:
You're welcome, I've already solved our dependency on dynamic plugins and am fine with continuing to use musl/mimalloc, but I'm sure it will help some people out.

Next up will be a small convenience PR for mimalloc.

With dynamic extension loading being possible, are there any that you'd like to offer out of the box or will this be up to package maintainers?

dunglas commented:
For convenience I think we should continue to package extensions statically as much as possible

DubbleClick commented:
> For convenience I think we should continue to package extensions statically as much as possible

I'm talking about extensions that cannot be compiled statically, such as xdebug, new relic (well, if they supported ZTS builds), or blackfire.

dunglas commented:
Got it. Maybe xdebug then. The other ones are proprietary.
